src/hardware/sblaster.cpp

```diff
@@ -1803,11 +1974,14 @@ class SBLASTER final : public Module_base {
 		case OPL_opl3:
 		case OPL_opl3gold:
 			OPL_Init(section,oplmode);
+			configure_opl_filter();
 			break;
 		}
 		if (sb.type==SBT_NONE || sb.type==SBT_GB) return;
 
 		sb.chan = MIXER_AddChannel(&SBLASTER_CallBack, 22050, "SB");
+		configure_sb_filter();
+
```

Comment:
- `CMS_Init()` creates the "CMS" or "GAMEBLASTER" channel (not named "FM"), so that's where its channel filter should be toggled (gameblaster.cpp:84)

 - Similarly, `OPL_Init()` is where the "FM" channel is created (adlib.cpp:256), and it's channel filter could be toggled there too (as it has access to the `[sblaster]` settings too.

Reply:
Okay, to make sure we're on the same page, I'll have these filter settings in the end:

```
sb_filter = auto|sb1|sb2|sbpro1|sbpro2|sb16|off
opl_filter = auto|sb1|sb2|sbpro1|sbpro2|sb16|off
cms_filter = on|off
```

So the CMS filter will be handled in `gameblaster.cpp`, yes.

`sb_filter` sets the filter of the SB PCM path, but there's no mentioning of the term "PCM" in the configs anywhere, so I figured it's best not to introduce that. I have to say, it's really really hard to make meaningful and consistent additions to the config, given that almost everything in it is a royal inconsistent mess... (not the DOSBox Staging teams fault, but still very frustrating).

By the way, I find the "shapeshifting" nature the of the CMS mixer channel (that it gets renamed to GAMEBLASTER when `sbtype=gb` is used) nonsensical, confusing, and totally random. The Game Blaster was just a rebranding of the Creative Music System; they literally put it in a different cardboard box and slapped a different logo on it, because the marketing guys thought it was a brilliant idea. But that's just a marketing manouvre; the CMS and the "Game Blaster" hardware are 100% identical!

"The [Game Blaster](http://vgmpf.com/Wiki/index.php?title=Game_Blaster) is just a C/MS with different packaging. The two products are physically identical." ([source](http://vgmpf.com/Wiki/index.php/Creative_Music_System))

So in my case, I most definitely don't want to have another `gb_filter` config param that is an alias to the `cms_filter` or something...

OPL does have the shapeshifting behaviour too: it's `opl*` in the configs, then it miraculously becomes "FM" in the mixer and 50% of the logs (I get it that the OPL is doing FM-synthesis, but we should use the same name for the same thing, so I'd just prefer OPL everywhere).

To add insult to injury, who ever thought this is a good idea: `oplmode=cms`. Huh? Why?!?! The CMS has _nothing_ to do with the Yamaha OPL chips. There should have been a `cms=on|off` option there from day one, completely separate from `oplmode`. Then if you specify `oplmode=cms`, you only get CMS, but no OPL. Okay. Then if you use `opl2` or `auto`, you get both CMS and OPL, but only if `sbtype` is `sb1` or `sb2`...

Anyway, rant off, I'm gonna drink some green tea now and play with the cats to calm down... 😹 🫖 

TL;DR I'll shoehorn the CMS filter into `gameblaster.cpp` somehow using the config param `cms_filter`. It should be totally unrelated to either the SB and the OPL.

Reply:
`configure_opl_filter()` applies its filters to the "FM" channel, but the GameBlaster and CMS channels are named differently, so these will slip past that "FM" channel lookup:

``` text
2022-05-28 18:07:47.389 | MIXER: CMS channel operating at 48000 Hz without resampling
2022-05-28 18:07:47.394 | CMS: Running on port 220h with two 7.159 MHz Phillips SAA-1099 chips

(no mention of filters running)
```

(or this could be out of scope, given these aren't strictly sound blaster.. and perhaps there are different audio goodies needed to treat these stream! -- just was flagging this in anycase)

Reply:
Yeah, this is intentional. The emulated Sound Blaster card sets the correct OPL filter for the "FM" channels, and only for that only -- that's correct, that's what I want here. Like I said, that's because we want to set the PCM and OPL filters per Sound Blaster model, but only for OPL, CMS is out of scope here (I've mapped `SBT_GB` to `FilterType::None` at line 321, but I only did that because I find it good practice to handle all possible enum values exhaustively; it's not really needed in actual practice).

So I expect to have a completely separate `configure_filter` method in `gameblaster.cpp` that just sets the fixed filter on or off.

Thanks for flagging, in any case. 👍🏻

Reply:
> sb_filter sets the filter of the SB PCM path, but there's no mentioning of the term "PCM" in the configs anywhere, so I figured it's best not to introduce that. I have to say, it's really really hard to make meaningful and consistent additions to the config, given that almost everything in it is a royal inconsistent mess... (not the DOSBox Staging teams fault, but still very frustrating). Gives me flashbacks of the PHP standard library (and that's not a good thing).

Indeed; DOSBox grew organically over time, so what might have started as a minimal configuration became the precedent as more features were added - but the style and names soon became constraining incumbents.

> By the way, I find the "shapeshifting" nature the of the CMS mixer channel (that it gets renamed to GAMEBLASTER when sbtype=gb is used) nonsensical, confusing, and totally random. 

Absolutely; I can see a bit of the appeal (GB being a precursor product); but to me, it really deserves its own settings, especially given it will soon have it's own filter setting too.  

> The Game Blaster was just a rebranding of the Creative Music System; they literally put it in a different cardboard box and slapped a different logo on it, because the marketing guys thought it was a brilliant idea. But that's just a marketing manouvre; the CMS and the "Game Blaster" hardware are 100% identical!

>    "The [Game Blaster](http://vgmpf.com/Wiki/index.php?title=Game_Blaster) is just a C/MS with different packaging. The two products are physically identical." ([source](http://vgmpf.com/Wiki/index.php/Creative_Music_System))

> So in my case, I most definitely don't want to have another gb_filter config param that is an alias to the cms_filter or something...

:+1: 

> To add insult to injury, who ever thought this is a good idea: oplmode=cms. Huh? Why?!?! The CMS has nothing to do with the Yamaha OPL chips.

`_dos_pc_mode = `amiga_500` 

Lol; indeed.  You can see some of this evolution in 74b2f10fe6f93cbfd1fc37e14b5d5b615d0d746e , back when separate CMS and Adlib conf settings were available.

> There should have been a cms=on|off option there from day one, completely separate from oplmode. Then if you specify oplmode=cms, you only get CMS, but no OPL. Okay. Then if you use opl2 or auto, you get both CMS and OPL, but only if sbtype is sb1 or sb2...

In that commit, the list was growing vertically allowing users to toggle a lot of stuff simultaneously: give me an Adlib, a SB16, and a CMS card,. collapsing things horizontally would have simplified the code and testing scenarios (so I can see one benefit; but the key name should have been umbrella'd up to the highest category (maybe `creative_card = [gb / sb1 / ..]`

> The original DOSBox authors reallly, really, really had dropped the ball on the whole OPL/CMS/FM/GAMEBLASTER fiasco, and now we're to live with this mess 🤦🏻‍♂️ face_with_spiral_eyes

I would argue that the benefit of fixing these confusing and incorrect settings outweighs maintaining their API compatibility, given these are very rarely used and very poorly understood by users. 

Plus.. of the rare users who are afficiandos of these early creative products, I expect they would  appreciate these being given more first-class billing instead of shoehorned in as afterthoughts. 

> Anyway, rant off, I'm gonna drink some green tea now and play with the cats to calm down... joy_cat teapot

Sounds very nice indeed :-)